Various improvements for the Linux input driver

jwrdegoede opened the pull request:
Various improvements for the Linux input driver:

- Allow opening the input device by device-name
- Re-acquire the device on connection loss
- Drop no longer needed custom mx5000 input handling

jwrdegoede committed:
server/drivers/linux_input: Add support for opening the device by name

The numbering of /dev/input/event# nodes depends on probe ordering and as
such is not stable.

Add support for opening an input-device by its name. If multiple devices
with the same name exists, the first one which matches will be used.

Signed-off-by: Hans de Goede <hdegoede@redhat.com>

jwrdegoede committed:
server/drivers/linux_input: Use strtol to get the key code

Use strtol to get the key code, so that it may also be specified as a
hexadecimal number.

Signed-off-by: Hans de Goede <hdegoede@redhat.com>

jwrdegoede committed:
server/drivers/linux_input: Re-acquire the device on connection loss

We may temporary loose access to the device. Possible causes are e.g.:
1. A temporary connection loss (Bluetooth); or
2. The device dropping of the bus to re-appear with another prod-id
   (the G510 keyboard does this when (un)plugging the headphones).

This commit adds support for trying to re-acquire the device if it was opened
by name, to deal with / workaround these temporary device losses.

Signed-off-by: Hans de Goede <hdegoede@redhat.com>

jwrdegoede committed:
server/drivers/mx5000: Drop custom input handling

There are kernel patches pending upstream which add proper evdev codes for
the enu buttons below the status LCD-s builtin to some keyboards.

Combined with recent improvements to LCDd's linux_input driver this means
we can use LCDd's linux_input driver instead of using custom input handling
in the mx5000 driver. Also the hardcoded scancodes used by the custom
handling are only correct for the mx5000 and not for the also supported
mx5500 model.

So lets remove this custom handling before we do a release which includes
this custom handling and people may start to rely on it.

Signed-off-by: Hans de Goede <hdegoede@redhat.com>

jwrdegoede committed:
server/drivers/linux_input: Only use defaults when there is no user config

Before this commit the linux_input event-code to event-string mapping
was hardcoded for ESC, Enter and the cursor-keys and there was no way
to override the defaults, or ignore these keys. Since the defaults where
checked before checking for user configured mappings.

This commit fixes this by first checking if there are user configured
mappings and if those are present then only using the user configured
mappings.

Signed-off-by: Hans de Goede <hdegoede@redhat.com>

jwrdegoede commented:
I've just pushed one more fix: "Only use the default mappings when there is no user config"

haraldg commented on server/drivers/linux_input.c:
```diff
@@ -183,6 +185,7 @@ linuxInput_init (Driver *drvthis)
 					drvthis->name, s);
 			return -1;
 		}
+		p->name = s;
```
I believe we should strdup() here. It probably works without, but I think we have no guarantee, that the server keeps the memory s is pointing at around.

Otherwise this PR looks perfect to me.

jwrdegoede replied:
Heh, I actually checked the lifetime of the string returned by config_get_string before doing things like this, the string is valid until config_clear() gets called.

The only caller of config_clear() is do_reload() from server/main.c which first calls drivers_unload_all() so the driver gets closed before the string returned from config_get_string() becomes invalid. IOW no need to strdup.

haraldg replied:
Yes, it works. But the point is: We have an API between server and drivers and ATM this API doesn't guarantee that the server will keep the string around. Of course we can just update the API in this regard. (I kind of would love to - I hate the amount of pointless strdup() we do for this reason.) So the real question is: Should we do this? I kind of understand the point between not mixing server allocated memory and driver allocated memory in the same data structure, so I never dared proposing something like that. But I'm certainly willing to be convinced.

However, actually the whole question isn't that important, because we probably will switch our configuration management soon after the next release to libelektra ...

jwrdegoede replied:
Ok, if this is going to change anyways I'm going to keep this as is. Let me know if I misunderstood and you still want me to add the strdup.

jwrdegoede committed:
server/drivers/linux_input: Consume all events until we find a valid keypress

Since we are only polling at framerate times per second, which with some
slow LCDs might be quite slow, we may build a backlog of events.

This is especially a problem when the input device we use also generates
a lot of events which are not for us, such as happens e.g. with the mx5000
and mx5500 keyboards where there is only 1 input device for both the
regular keys and the lcd-menu buttons.

This can cause a huge amount of lag, this commit fixes this by consuming
all events until we find a valid keypress; or there are no more events.

Signed-off-by: Hans de Goede <hdegoede@redhat.com>

jwrdegoede commented:
I've just pushed one more fix: "server/drivers/linux_input: Consume all events until we find a valid keypress"

haraldg commented on server/drivers/linux_input.c:
```diff
@@ -288,20 +286,35 @@ linuxInput_get_key (Driver *drvthis)
+ *               \c NULL for nothing available / error.
+ */
+const char *
+linuxInput_key_code_to_key_name (PrivateData *p, uint16_t code)
```
Make this function static?

jwrdegoede replied:
Ack, good one, I will push a new version with this fixed in a minute.

jwrdegoede commented:
As promised I've just pushed a new version with the missing static added.
